### ISSUE TITLE
Question about training custom dataset

### ISSUE THREAD

**Issue by lohzhunyewcs** [ying09/TextFuseNet#11]
Hi I have 2 questions about training,

1) Are both bbox and segmentation used? 
2) Do we have to label character by character or would word/sentences be fine as well? In the sample json it does seem to have labels only by character( e.g.annotation[0]= category_id: 1, segmentation:[[...]]) but since it's a detection model I'm not sure if the label(category_id) matters.

Thanks for your time and the model!

**Comment by ying09** [comment]
1. Yes
2. We need to annotate characters and words to training TextFuseNet.
Please refer to ./TextFuseNet/annotation_example/ carefully.

**Comment by lohzhunyewcs** [comment]
Okay so just to confirm, category_id for 1 is the word and the rest are for each character right?

**Comment by ying09** [comment]
1. Yes
2. After pre-training on SynthText, we apply the pre-trained model on IC15, Total-Text and CTW-1500 to search for character training samples and save them in a txt file. After that, we make the training json file in C++. I think you can do this easily in Python.

**Comment by lohzhunyewcs** [comment]
I see, may I know what pre-trained model you guys used?

**Comment by Real-YeJ** [comment]
@lohzhunyewcs  I suggest that you use our pre-trained totaltext-r101 model, which can detect arbitrary texts. You need to modify the code of totaltext_detection.py to detect characters for annotations. or you can train a pre-trained model according to weakly supervised learning in our paper.

**Comment by lohzhunyewcs** [comment]
@Real-YeJ I see thanks for the suggestion! Since the ctw model outperforms the totaltext for my current data set, would it be more wise to use that instead? As for the modification, did you mean changing the classes[i]==0 to != 0 since 0 is for full text whereas the rest are characters right? Also if that's the case, wouldn't it be possible for textfusenet to perform recognition since it detects characters?

**Comment by Real-YeJ** [comment]
@lohzhunyewcs  yeah, 0 class is word-level text, and !0 is character-level text. The performance of totaltext and ctw-1500 is  similar.  As you said, TextFuseNet can realize text recognition through displaying classification. But this is only for English text, which does not work when detecting other text, such as Chinese. We're doing new research about mutil-language text. In short, What we want to express is that detecting characters and extract global-level features can enhance the network's perception for text.

**Comment by lohzhunyewcs** [comment]
@Real-YeJ I see that's an interesting research. Is it possible that the reason for the poor recognition for chinese text is due to the large amount of possible words hence requiring a much larger dataset/more complex model? Or is it because it is harder to distinguish whether words such as 双 can be easily mistaken as 2 又?

On a separate note, I've tried both models on a document based dataset.

totaltext
- Image size(min, max) = 900
- Managed to detect some(<20%) text and characters(all accurate recognition wise)

CTW-1500
- Image size(min, max) = 990
- Detected 98+% of texts but no characters detected at all.

I'm not so sure how to proceed from this because while CTW-1500 managed to detect most of the texts that I need, it doesn't detect any characters at all. On the other hand, totaltext doesn't really detect many text nor sufficiently enough characters. I would still like to fine tune the model however, for some of the texts the CTW-1500 missed out on and since most of my text should be consisting of just alphabets, it'd be nice if it could help perform recognition as well. 

Would cropping out the images into smaller texts then running the pre-trained model be better since as far as I'm aware of totaltext and ctw are mostly outdoor images or text on real world objects.

**Comment by Real-YeJ** [comment]
@lohzhunyewcs    
actually, in this implementation, we train ctw-1500 via only fine-tuning totaltext-model on original images and annotations of ctw-1500, which shows outstanding result as well. generating weakly supervised char-level lables and training with three levels of features are better. so I suggest that you use totaltext-model to get characters.  

The document based dataset has domain shift with scene text dataset, which may cause the problem you mentioned.  if the model based on scene text datasets does not work, I think you can generate synthetic document datasets(focus on https://github.com/ankush-me/SynthText ), pretrain a model, and then use the pre-trained model to detect characters for real document based dataset.

**Comment by lohzhunyewcs** [comment]
@Real-YeJ Oh I see, thanks for the help then, I'll try and see how it goes

**Comment by thanhhau097** [comment]
@lohzhunyewcs could you please tell me how to convert the word/sentence level annotations to the annotations needed to train the TextFuseNet? I've read the annotation examples and comments, but I am sure that I fully understanded them. It's better if I have a sample code. Thank you.

**Comment by lohzhunyewcs** [comment]
@thanhhau097 they recommend training a pretrained model like the totaltext one they have on character level annotations like with Synthtext then use it to detect the characters in your dataset to get the annotations.

**Comment by kbrajwani** [comment]
@Real-YeJ  Thanks for such a good research project.
> actually, in this implementation, we train ctw-1500 via only fine-tuning totaltext-model on original images and annotations of ctw-1500, which shows outstanding result as well. 
By this you mean you have used ctw-1500 annotation with only text or text and character level?

I am currently training on https://guillaumejaume.github.io/FUNSD/download/ this dataset. which is based on documents where word is annotated. i have modified your code little bit to train only on word level. 

Currently i am training with pretrained ctw model. Following is my metrics file can you tell me its looks fine or i have done something wrong?
[metrics.txt](https://github.com/ying09/TextFuseNet/files/5790243/metrics.txt)
